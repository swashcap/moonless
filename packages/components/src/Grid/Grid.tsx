import React from 'react';
import { clsx } from 'moonless-utils';

import styles from './Grid.module.css';

export type GridProps = React.DetailedHTMLProps<
  React.HTMLAttributes<HTMLDivElement>,
  HTMLDivElement
>;

export type GridColSpan =
  | '1'
  | '2'
  | '3'
  | '4'
  | '5'
  | '6'
  | '7'
  | '8'
  | '9'
  | '10'
  | '11'
  | '12';

export interface GridColProps
  extends React.DetailedHTMLProps<
    React.HTMLAttributes<HTMLDivElement>,
    HTMLDivElement
  > {
  span?: GridColSpan;
}

export class Grid extends React.Component<GridProps> {
  static Col({ className, span, ...rest }: GridColProps) {
    return (
      <div
        className={clsx(
          styles.GridCol,
          span === '1' && styles.GridCol1,
          span === '2' && styles.GridCol2,
          span === '3' && styles.GridCol3,
          span === '4' && styles.GridCol4,
          span === '5' && styles.GridCol5,
          span === '6' && styles.GridCol6,
          span === '7' && styles.GridCol7,
          span === '8' && styles.GridCol8,
          span === '9' && styles.GridCol9,
          span === '10' && styles.GridCol10,
          span === '11' && styles.GridCol11,
          span === '12' && styles.GridCol12,

          className
        )}
        {...rest}
      />
    );
  }

  render() {
    const { className, ...rest } = this.props;

    return <div className={clsx(styles.Grid, className)} {...rest} />;
  }
}
