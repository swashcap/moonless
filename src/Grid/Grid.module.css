/** Grid */

.Grid {
  display: flex;
  flex-wrap: wrap;
  margin-left: calc(-1 * var(--space-100));
  margin-right: calc(-1 * var(--space-100));
}

/** Grid: Col */

.GridCol {
  flex: 1;
  padding-left: var(--space-100);
  padding-right: var(--space-100);
}

.GridCol1 {
  width: 8.333%;
}

.GridCol2 {
  width: 16.66%;
}

.GridCol3 {
  width: 25%;
}

.GridCol4 {
  width: 33.33%;
}

.GridCol5 {
  width: 41.66%;
}

.GridCol6 {
  width: 50%;
}

.GridCol7 {
  width: 58.33%;
}

.GridCol8 {
  width: 66.66%;
}

.GridCol9 {
  width: 75%;
}

.GridCol10 {
  width: 83.33%;
}

.GridCol11 {
  width: 91.66%;
}

.GridCol12 {
  width: 100%;
}
