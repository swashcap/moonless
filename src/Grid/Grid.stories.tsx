import React from 'react';

import { Grid } from './Grid';

export default {
  component: Grid,
  title: 'Grid',
};

export const Default = () => (
  <Grid>
    <Grid.Col>1</Grid.Col>
    <Grid.Col>2</Grid.Col>
    <Grid.Col>3</Grid.Col>
  </Grid>
);
