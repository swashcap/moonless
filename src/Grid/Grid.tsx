import React from 'react';

import { clsx } from '../utils/clsx';
import styles from './Grid.module.css';

export type GridProps = React.DetailedHTMLProps<
  React.HTMLAttributes<HTMLDivElement>,
  HTMLDivElement
>;

export type GridColProps = React.DetailedHTMLProps<
  React.HTMLAttributes<HTMLDivElement>,
  HTMLDivElement
>;

export class Grid extends React.Component<GridProps> {
  static Col({ className, ...rest }: GridColProps) {
    return <div className={clsx(styles.GridCol, className)} {...rest} />;
  }

  render() {
    const { className, ...rest } = this.props;

    return <div className={clsx(styles.Grid, className)} {...rest} />;
  }
}
