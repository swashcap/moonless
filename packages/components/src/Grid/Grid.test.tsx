import { render } from '@testing-library/react';

import { Grid, GridColSpan } from './Grid';

describe('Grid', () => {
  test('It should render.', () => {
    const { container } = render(<Grid />);

    expect(container).toMatchSnapshot();
  });

  test('It should pass className property.', () => {
    const { getByTestId } = render(
      <Grid className="test" data-testid="grid" />
    );

    expect(getByTestId('grid')).toHaveClass('test');
  });

  test('It should spread properties to the root element.', () => {
    const { getByTestId } = render(
      <Grid data-greetings="hello" data-sup="hi" data-testid="grid" />
    );
    const grid = getByTestId('grid');

    expect(grid).toHaveAttribute('data-greetings', 'hello');
    expect(grid).toHaveAttribute('data-sup', 'hi');
  });
});

describe('GridCol', () => {
  test('It should render.', () => {
    const { container } = render(<Grid.Col />);

    expect(container).toMatchSnapshot();
  });

  test('It should pass className property.', () => {
    const { getByTestId } = render(
      <Grid.Col className="test" data-testid="grid-col" />
    );

    expect(getByTestId('grid-col')).toHaveClass('test');
  });

  test('It should spread properties to the root element.', () => {
    const { getByTestId } = render(
      <Grid.Col data-greetings="hello" data-sup="hi" data-testid="grid-col" />
    );
    const grid = getByTestId('grid-col');

    expect(grid).toHaveAttribute('data-greetings', 'hello');
    expect(grid).toHaveAttribute('data-sup', 'hi');
  });

  test.each<GridColSpan>([
    '1',
    '2',
    '3',
    '4',
    '5',
    '6',
    '7',
    '8',
    '9',
    '10',
    '11',
    '12',
  ])('It should render column span: %s.', (span) => {
    const { getByTestId } = render(
      <Grid.Col data-testid="grid-col" span={span} />
    );

    expect(getByTestId('grid-col')).toHaveClass(`GridCol${span}`);
  });
});
