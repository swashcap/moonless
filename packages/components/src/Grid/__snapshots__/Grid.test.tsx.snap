// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid It should render. 1`] = `
<div>
  <div
    class="Grid"
  />
</div>
`;

exports[`GridCol It should render. 1`] = `
<div>
  <div
    class="GridCol"
  />
</div>
`;
